onnxruntime/core/optimizer/group_query_attention_fusion.cc

```diff
@@ -0,0 +1,478 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/group_query_attention_fusion.h"
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/utils.h"
+
+#define DEBUG_LOG(x) LOGS(logger, VERBOSE) << x
+
+using namespace ONNX_NAMESPACE;
+namespace onnxruntime {
+
+static void AttachNodeAttribute(Node& node, const std::string& attribute_name, int64_t attribute_value, AttributeProto_AttributeType attribute_type) {
+  NodeAttributes& node_attributes = node.GetMutableAttributes();
+  AttributeProto attr;
+  attr.set_name(attribute_name);
+  attr.set_i(attribute_value);
+  attr.set_type(attribute_type);
+  node_attributes[attribute_name] = attr;
+}
+
+static NodeArg& MergeQkvWeightsForMatMul(Graph& graph,
+                                         int64_t q_hidden_size,
+                                         int64_t kv_hidden_size,
+                                         const TensorProto* q_tensor,
+                                         const TensorProto* k_tensor,
+                                         const TensorProto* v_tensor) {
+  assert(nullptr != q_tensor);
+  assert(nullptr != k_tensor);
+  assert(nullptr != v_tensor);
+
+  int64_t input_hidden_size = q_tensor->dims(0);
+
+  Initializer q_initializer(*q_tensor, graph.ModelPath());
+  Initializer k_initializer(*k_tensor, graph.ModelPath());
+  Initializer v_initializer(*v_tensor, graph.ModelPath());
+
+  int64_t output_hidden_size = q_hidden_size + 2 * kv_hidden_size;
+
+  TensorProto qkv_b_initializer;
+  qkv_b_initializer.set_name(graph.GenerateNodeArgName("qkv_B"));
+  qkv_b_initializer.add_dims(input_hidden_size);
+  qkv_b_initializer.add_dims(output_hidden_size);
+  qkv_b_initializer.set_data_type(q_tensor->data_type());
+
+  const MLFloat16* q_data = q_initializer.data<MLFloat16>();
```

Comment:
This assumes that input is float16. Shall we check before the fusion that inputs are MLFloat16?